### PR TITLE
47: add validation for edit note and remove unneeded lookups for config changes

### PR DESCRIPTION
**Pull request by jshvarts**
Behavior in action can be viewed here: https://drive.google.com/open?id=0B-5eHbqh90ciaXFYMTV6blM1b1U

### CODE REVIEW

File: app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt [hunk 1]
```diff
@@ -30,6 +30,24 @@ class EditNotePresenter(private val view: EditNoteView,
...
+
+    fun setupTextViewsChanged() {
+        view.titleTextChanges()
+                .map { model.validateNoteTitleText(view.getNewNoteTitle()) }
+                .doOnNext { if (!it) { view.showInvalidNoteTitleError(); view.toggleSubmit(false) } }
+                // filters out anything that doesn't match the predicate (only emit true events (the text is valid).
+                .filter { it }
+                .subscribe { view.toggleSubmit(validateAllFields()) }
+
+        view.bodyTextChanges()
```
Comment (exallium) [inline app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt]: The only thing I think is really missing from kotlin is proper function references, like `model::validateNoteTitleText`, which'd really let you minimize and generify this method.

There are probably ways to get rid of the duplication here, but nothing exceedingly straightforward.

File: app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt [hunk 2]
```diff
@@ -30,6 +30,24 @@ class EditNotePresenter(private val view: EditNoteView,
         setupSaveNoteSubscription(noteId)
 
         view.cancelEditNoteClicks().map { null }.subscribe(showNewNoteDetailSubscriberFn).addToComposite()
+
+        setupTextViewsChanged()
+    }
+
+    fun setupTextViewsChanged() {
+        view.titleTextChanges()
+                .map { model.validateNoteTitleText(view.getNewNoteTitle()) }
```
Comment (exallium) [inline app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt]: what is the value of `it` in this lambda?

File: app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt [hunk 3]
```diff
@@ -30,6 +30,24 @@ class EditNotePresenter(private val view: EditNoteView,
...
+                .doOnNext { if (!it) { view.showInvalidNoteTitleError(); view.toggleSubmit(false) } }
+                // filters out anything that doesn't match the predicate (only emit true events (the text is valid).
+                .filter { it }
+                .subscribe { view.toggleSubmit(validateAllFields()) }
+
+        view.bodyTextChanges()
+                .map { model.validateNoteBodyText(view.getNewNoteBody()) }
```
Comment (exallium) [inline app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt]: what is the value of `it` in this lambda?

File: build.gradle [hunk 4]
```diff
@@ -6,7 +6,7 @@ buildscript {
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.3.0-beta3'
+        classpath 'com.android.tools.build:gradle:2.3.0'
```
Comment (exallium) [inline build.gradle:9]: nice.

**Review by exallium — CHANGES_REQUESTED** [review]
A few things I'd like feedback on

File: app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt (hunk 3 above)
Comment (jshvarts) [reply app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt]: done

File: app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt (hunk 2 above)
Comment (jshvarts) [reply app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt]: done

File: app/src/main/kotlin/com/exallium/todoapp/editnote/EditNoteViewImpl.kt [hunk 5]
```diff
@@ -68,6 +73,18 @@ class EditNoteViewImpl(val bundle: Bundle) : BaseViewImpl<EditNoteView, EditNote
         displaySnackbar(R.string.unable_to_save_note_error)
     }
 
+    override fun showInvalidNoteTitleError() {
+        noteTitleEditText.error = resources!!.getString(R.string.invalid_note_title_error)
```
Comment (exallium) [inline app/src/main/kotlin/com/exallium/todoapp/editnote/EditNoteViewImpl.kt]: `@BindString` ?

File: app/src/main/kotlin/com/exallium/todoapp/editnote/EditNoteViewImpl.kt [hunk 6]
```diff
@@ -68,6 +73,18 @@ class EditNoteViewImpl(val bundle: Bundle) : BaseViewImpl<EditNoteView, EditNote
...
+    }
+
+    override fun showInvalidNoteBodyError() {
+        noteBodyTextView.error = resources!!.getString(R.string.invalid_note_body_error)
```
Comment (exallium) [inline app/src/main/kotlin/com/exallium/todoapp/editnote/EditNoteViewImpl.kt]: `@BindString` ?

File: app/src/test/kotlin/com/exallium/todoapp/editnote/EditNotePresenterTest.kt [hunk 7]
```diff
@@ -73,7 +70,7 @@ class EditNotePresenterTest {
         // THEN
         verify(screenBundleHelper).getNoteId(bundle)
     }
-
+
```
Comment (exallium) [inline app/src/test/kotlin/com/exallium/todoapp/editnote/EditNotePresenterTest.kt:73]: this seems like a weird spot for spaces

**Review by exallium — CHANGES_REQUESTED** [review]
I think we want to use `BindString`. good otherwise.

File: app/src/main/kotlin/com/exallium/todoapp/editnote/EditNoteViewImpl.kt (hunk 5 above)
Comment (jshvarts) [reply app/src/main/kotlin/com/exallium/todoapp/editnote/EditNoteViewImpl.kt]: done

File: app/src/test/kotlin/com/exallium/todoapp/editnote/EditNotePresenterTest.kt (hunk 7 above)
Comment (jshvarts) [reply app/src/test/kotlin/com/exallium/todoapp/editnote/EditNotePresenterTest.kt:73]: github complained http://stackoverflow.com/questions/5813311/no-newline-at-end-of-file

File: app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt (hunk 1 above)
Comment (jshvarts) [reply app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt]: could you make a follow-up commit with how you'd improve it so I can learn from it?

File: app/src/test/kotlin/com/exallium/todoapp/editnote/EditNotePresenterTest.kt (hunk 7 above)
Comment (exallium) [reply app/src/test/kotlin/com/exallium/todoapp/editnote/EditNotePresenterTest.kt:73]: =) ok

File: app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt (hunk 1 above)
Comment (exallium) [reply app/src/main/kotlin/com/exallium/todoapp/editnote/EditNotePresenter.kt]: I'll do a commit which generifies this and let you see it, yeah.

**Review by exallium — APPROVED** [review]

**Pull request merged**